Daniel - Update DG and PPP (#203)
AY2122S2-CS2103-W17-1/tp: docs/DeveloperGuide.md

</div>

> Within the `/model` folder, there also exists an `IndexedCsvFile` model that helps with the parsing of CSV files for the Import CSV function. However, this file does not fit within the model component diagram and just serves as a helper model. 

Reviewer: Please avoid calling it a folder.
```suggestion
Within the `model` package, there also exists an `IndexedCsvFile` model that helps with the parsing of CSV files for the Import CSV function. However, the class does not maintain any persistent instances, and does not fit within the model component diagram, serving solely as a helper model. 
```

Author: thanks for the suggestion, implemented in https://github.com/AY2122S2-CS2103-W17-1/tp/pull/203/commits/3193b02116aeddfa74a438d365ea5575d9acfb0b

* inherits from both `AddressBookStorage` and `UserPrefStorage`, which means it can be treated as either one (if only the functionality of only one is needed).
* depends on some classes in the `Model` component (because the `Storage` component's job is to save/retrieve objects that belong to the `Model`)

> Within the `/storage` folder, there also exists a `CsvManager` class that is a helper file to do file IO related operations when Import CSV or Export CSV is called. This file does not fit within the storage component diagram as it is simply meant to separate IO operations to a helper class within `/storage`.

Reviewer: Likewise, please call it a package.
```suggestion
Within the `storage` package, there also exists a `CsvManager` class that is a helper that provides file IO related operations for working with CSV files. It is used in the Import CSV or Export CSV features. This class does not fit within the storage component diagram as it does not interact with any other classes, serving solely as a helper class.
```

Author: thanks for the suggestion, implemented in https://github.com/AY2122S2-CS2103-W17-1/tp/pull/203/commits/3193b02116aeddfa74a438d365ea5575d9acfb0b

Reviewer: I think you might have missed this.

Author: oops yes i missed this, thanks! resolved in https://github.com/AY2122S2-CS2103-W17-1/tp/pull/203/commits/805457bc58e7b818e04fe9ea0926b906040f4fd5

This section describes some noteworthy details on how certain features are implemented.

### Import and Export CSV Features

This section describes some of the details as to how the import and export CSV features were implemented

#### Import CSV

> The import CSV function is meant to append to the current address book with new data imported from any CSV file. The intention is to allow users to be able to import from a Microsoft Excel compatible format. Since there are multiple different templates for contacts in CSV files across various platforms, such as Microsoft Outlook and Google Contacts, the idea was to ensure this feature was as flexible as possible, allowing the user to specify which particular columns contained certain information

Reviewer: I would suggest a change in phrasing here.
```suggestion
The import CSV function is meant to append to the current address book with new data imported from any CSV file. The intention is to allow users to be able to import from a Microsoft Excel compatible format. Since there are multiple different templates for contacts in CSV files across various platforms, such as Microsoft Outlook and Google Contacts, the feature is designed to be as flexible as possible, allowing the user to specify mappings for the information contained in the various columns.
```

Author: thanks for the suggestion! implemented in https://github.com/AY2122S2-CS2103-W17-1/tp/pull/203/commits/3193b02116aeddfa74a438d365ea5575d9acfb0b

The arguments that are parsed here are the custom column numbers for each value, e.g. `n/3 p/4 e/5 a/6 t/7` will read `Name` from column 3, `Phone` from column 4, `Email` from column 5, `Address` from column 6 and `Tags` from column 7 

> In the event that any of the data read does not conform to the restrictions given by each of the components in `Person`, e.g. an email that does not have `@`, or if the person has a duplicate name with another person, that particular line will be skipped.

Reviewer: I suggest deferring the example to the end after you have completed your sentence.

```suggestion
In the event that any of the data fields read do not conform to the restrictions given by each of the components in `Person`, that particular line will be skipped. For example, if the record in a line has an email that does not have the `@` symbol or if the record contains a duplicated name that already exists in the Address Book, the entire line will be skipped.
```

Author: thanks for the suggestion, applied in https://github.com/AY2122S2-CS2103-W17-1/tp/pull/203/commits/3193b02116aeddfa74a438d365ea5575d9acfb0b

> The `CSVManager` takes in the `IndexedCsvFile` object, opens the file and reads the lines, while a function from the command is passed to the Manager that will store each of the lines as a new Person in the model. The sequence diagram is as follows:

Reviewer: I suggest this version of the description, which better captures where the logic is.
```suggestion
The `CSVManager` takes in a `IndexedCsvFile` object, opens the file and reads the lines. The logic that performs the parsing of data fields and creation of `Person` models is specified in the import command, and is passed as an anonymous function to CSVManager. The sequence diagram is as follows:
```

Author: thanks for the suggesiton! implemented in https://github.com/AY2122S2-CS2103-W17-1/tp/pull/203/commits/3193b02116aeddfa74a438d365ea5575d9acfb0b

![ImportCsvSequenceDiagram](images/ImportCsvSequenceDiagram.png)

#### Export CSV

The Export CSV function is meant to value add upon the existing `.json` file saving to provide an alternative option, especially for users who prefer interacting with a Microsoft Excel compatible format.

The resulting `addressbook.csv` that will be produced will be in the following format

| Name        | Phone        | Email        | Address        | Tags                            |
|-------------|--------------|--------------|----------------|---------------------------------|
| Person Name | Person Phone | Person Email | Person Address | Person Tag 1;Person Tag 2; .... |

> Multiple tags are delimited by the `;` character

Reviewer: Missing sentence terminator.
```suggestion
Multiple tags are delimited by the `;` character.
```

Author: thanks for the spot, implemented in https://github.com/AY2122S2-CS2103-W17-1/tp/pull/203/commits/3193b02116aeddfa74a438d365ea5575d9acfb0b

> Likewise with Import, File IO operations are also separated into the `CSVManager` class, such that the command logic takes the current address book, parses them into the approprate strings, then passes it to the `CsvManager` to handle the writing to file

Reviewer: This phrasing is very strange.
```suggestion
Like the Import feature, File IO operations are also separated into the `CSVManager` class, such that the command logic takes the current address book, parses them into the appropriate strings, then passes it to the `CsvManager` to handle the writing to file.
```

Author: thanks for the suggestion, implemented in https://github.com/AY2122S2-CS2103-W17-1/tp/pull/203/commits/3193b02116aeddfa74a438d365ea5575d9acfb0b

The sequence diagram is as follows:
![ExportCsvSequenceDiagram](images/ExportCsvSequenceDiagram.png)

> The subsequent exported file can be imported back into any other instance of ContaX, similar to the existing `.json` system of import/export

Reviewer: Missing Fullstop.
```suggestion
The exported file can be subsequently imported back into any other instance of ContaX, similar to the existing `.json` system of import/export.
```

Author: thanks for the catch, implemented in https://github.com/AY2122S2-CS2103-W17-1/tp/pull/203/commits/3193b02116aeddfa74a438d365ea5575d9acfb0b
